ssz/src/main/java/tech/pegasys/artemis/util/backing/view/AbstractCompositeViewRead.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.backing.view;
+
+import tech.pegasys.artemis.util.backing.CompositeViewRead;
+import tech.pegasys.artemis.util.backing.ViewRead;
+import tech.pegasys.artemis.util.backing.cache.ArrayCache;
+import tech.pegasys.artemis.util.backing.cache.IntCache;
+import tech.pegasys.artemis.util.backing.tree.TreeNode;
+import tech.pegasys.artemis.util.backing.type.CompositeViewType;
+
+/**
+ * Base backing view class for immutable composite views (lists, vectors, containers)
+ *
+ * <p>It caches it's child view instances so that if the underlying tree nodes are not changed (in
+ * the corresponding mutable classes) the instances are not recreated from tree nodes on later
+ * access.
+ *
+ * <p>Thought internally this class has a mutable cache it may be thought of as immutable instance
+ * and used safely across threads
+ *
+ * @param <ChildType> the type of children. For heterogeneous composites (like container) this type
+ *     would be just generic {@link ViewRead}
+ */
+public abstract class AbstractCompositeViewRead<ChildType extends ViewRead>
+    implements CompositeViewRead<ChildType> {
+
+  private IntCache<ChildType> childrenViewCache;
+  private final int sizeCache;
+  private final CompositeViewType type;
+  private final TreeNode backingNode;
+
+  /** Creates an instance from a type and a backing node */
+  protected AbstractCompositeViewRead(CompositeViewType type, TreeNode backingNode) {
+    this(type, backingNode, null);
+  }
+
+  /**
+   * Creates an instance from a type and a backing node Optionally the view instances cache can be
+   * supplied for optimization to shortcut children views creation from backing nodes. The cache
+   * should correspond to the supplied backing tree.
+   */
+  protected AbstractCompositeViewRead(
+      CompositeViewType type, TreeNode backingNode, IntCache<ChildType> cache) {
+    this.type = type;
+    this.backingNode = backingNode;
+    sizeCache = sizeImpl();
+    childrenViewCache = cache == null ? createCache() : cache;
+  }
+
+  /**
+   * 'Transfers' the cache to a new Cache instance eliminating all the cached values from the
+   * current view cache. This is made under assumption that the view instance this cache is
```

Comment:
I just wanted to emphasize here that transferring the cache (not copying it) removes all the cached entries from this view's cache. This is actually how `Cache.transfer` is intended to work. 
Please correct my wordings here if the idea was expressed too unclear.

Reply:
No - that makes sense.  I just couldn't find where it was being cleared out because I was looking at the default impl in [`IntArray`](https://github.com/Nashatyrev/artemis/blob/af118aa35f2e3b92e303db06f717b66595fb75f8/ssz/src/main/java/tech/pegasys/artemis/util/backing/cache/IntCache.java#L44-L46).  But I see that the [child `ArrayIntCache`](https://github.com/Nashatyrev/artemis/blob/af118aa35f2e3b92e303db06f717b66595fb75f8/ssz/src/main/java/tech/pegasys/artemis/util/backing/cache/ArrayIntCache.java#L77-L81) implementation works as described here.  

Maybe we should cut the default implementation?

Reply:
I would again leave this default implementation for potential interface implementations which don't bother with optimizing `transfer()` - `copy()` is just a naive `transfer()`